exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/entities/EntityTests.kt

```diff
@@ -884,6 +884,73 @@ class EntityTests : DatabaseTestsBase() {
         }
     }
 
+    @Test
+    fun testIterationOverSizedIterableWithPreload() {
+        fun HashMap<String, Pair<Int, Long>>.assertEachQueryExecutedOnlyOnce() {
+            forEach { (statement, stats) ->
+                val executionCount = stats.first
+                assertEquals(1, executionCount, "Statement executed more than once: $statement")
+            }
+        }
+
+        withTables(Regions, Schools) {
+            val region1 = Region.new {
+                name = "United Kingdom"
+            }
+            School.new {
+                name = "Eton"
+                region = region1
+            }
+            School.new {
+                name = "Harrow"
+                region = region1
+            }
+
+            commit()
+
+            inTopLevelTransaction(Connection.TRANSACTION_SERIALIZABLE) {
+                debug = true // enables tracking of executed statements in this transaction
+
+                val allSchools = School.all().with(School::region).toList()
+
+                assertEquals(2, allSchools.size)
+                // expected: 1 query to select all School, and 1 query to select referenced Regions
+                assertEquals(2, statementCount)
+                assertEquals(statementCount, statementStats.size)
+                statementStats.assertEachQueryExecutedOnlyOnce()
+
+                // reset tracker
+                statementCount = 0
+                statementStats.clear()
+
+                val oneSchool = School.all().limit(1).with(School::region).toList()
+
+                assertEquals(1, oneSchool.size)
+                assertEquals(2, statementCount)
+                assertEquals(statementCount, statementStats.size)
+                statementStats.assertEachQueryExecutedOnlyOnce()
+
+                debug = false
+            }
+
+            // test that cached result doesn't propagate when SizedIterable query changes after loading
+            inTopLevelTransaction(Connection.TRANSACTION_SERIALIZABLE) {
+                debug = true
+
+                val oneSchool = School.all().with(School::region).limit(1).toList()
+
+                assertEquals(1, oneSchool.size)
+                // expected: 1 query to select all School, 1 query to select the referenced Regions,
+                // then 1 new query to select only first School
+                assertEquals(3, statementCount)
+                assertEquals(statementCount, statementStats.size)
+                statementStats.assertEachQueryExecutedOnlyOnce()
+
+                debug = false
```

Comment:
The current set up of the anonymous `SizedIterable` is that it makes new copies of the invoking source iterable (without the cached result) when adjusting the underlying query with chained functions.
So even when a result has been cached, chaining functions afterwards changes the expected query and triggers a new SQL execution on iteration because `loadedResult` is reset to `null`.

If it turns out that this is not the expected behavior for users, `loadedResult` logic can be easily added to the overrides in the anonymous function.